lib/attr_extras/attr_initialize/attributes.rb

```diff
@@ -0,0 +1,36 @@
+module AttrExtras
+  class AttrInitialize
+    class Attributes
+      def initialize(names_with_array_of_hash_arguments)
+        @names_with_array_of_hash_arguments = names_with_array_of_hash_arguments
+      end
+
+      def positional
+        @attributes_names ||= names_with_array_of_hash_arguments.take_while { |name_or_array| !name_or_array.is_a?(Array) }
+      end
+
+      def hash
+        @hash_names ||= hash_names_without_default_values + default_values.keys
+      end
+
+      def default_values
+        return @default_values if @default_values
+
+        @default_values = (hash_names_with_default_values - hash_names_without_default_values).first
+        @default_values ||= {}
+      end
+
+      private
+
+      def hash_names_with_default_values
+        (names_with_array_of_hash_arguments - positional).flatten
+      end
+
+      def hash_names_without_default_values
+        hash_names_with_default_values.take_while { |name| !name.is_a?(Hash) }
+      end
+
+      attr_reader :names_with_array_of_hash_arguments
```

Comment:
`private; attr_reader` gives warnings as mentioned [in the README](https://github.com/barsoom/attr_extras/#why-not-use-private-attr_reader-foo-instead-of-attr_private-foo) :D